prov/efa/src/rxr/rxr_pkt_cmd.c

```diff
@@ -325,77 +320,91 @@ ssize_t rxr_pkt_post_ctrl_once(struct rxr_ep *rxr_ep, int entry_type, void *x_en
- * @param[in]   x_entry         x_entry pointer
- * @param[in]   ctrl_type       type of control packet
+ * @param[in]   x_entry         pointer to rxr_op_entry. (either a tx_entry or an rx_entry)
+ * @param[in]   pkt_type        packet type.
  * @param[in]   inject          send control packet via inject or not.
- * @param[in]   flags           additional flags to apply for fi_sendmsg.
- *                              currently only accepted flags is FI_MORE.
- * @return      On success return 0, otherwise return a negative error code
+ * @return      On success return 0, otherwise return a negative libfabric error code. Typical error:
  */
-ssize_t rxr_pkt_post_ctrl(struct rxr_ep *ep, int entry_type, void *x_entry,
-			  int ctrl_type, bool inject, uint64_t flags)
+ssize_t rxr_pkt_post_or_queue(struct rxr_ep *ep, struct rxr_op_entry *op_entry, int pkt_type, bool inject)
 {
 	ssize_t err;
-	struct rxr_tx_entry *tx_entry;
-
-	if (ctrl_type == RXR_MEDIUM_TAGRTM_PKT ||
-	    ctrl_type == RXR_MEDIUM_MSGRTM_PKT ||
-	    ctrl_type == RXR_DC_MEDIUM_MSGRTM_PKT ||
-	    ctrl_type == RXR_DC_MEDIUM_TAGRTM_PKT) {
-		assert(entry_type == RXR_TX_ENTRY);
-		assert(!inject);
 
-		tx_entry = (struct rxr_tx_entry *)x_entry;
-		while (tx_entry->bytes_sent < tx_entry->total_len) {
-			err = rxr_pkt_post_ctrl_once(ep, RXR_TX_ENTRY, x_entry, ctrl_type, 0, flags);
-			if (OFI_UNLIKELY(err))
-				return err;
+	err = rxr_pkt_post(ep, op_entry, pkt_type, inject, 0);
+	if (err == -FI_EAGAIN) {
+		if (op_entry->type == RXR_TX_ENTRY) {
+			assert(!(op_entry->rxr_flags & RXR_TX_ENTRY_QUEUED_RNR));
+			op_entry->state = RXR_TX_QUEUED_CTRL;
+			op_entry->queued_ctrl.type = pkt_type;
+			op_entry->queued_ctrl.inject = inject;
+			dlist_insert_tail(&op_entry->queued_ctrl_entry,
+					  &ep->tx_entry_queued_ctrl_list);
+		} else {
+			assert(op_entry->type == RXR_RX_ENTRY);
+			assert(op_entry->state != RXR_RX_QUEUED_CTRL);
+			op_entry->state = RXR_RX_QUEUED_CTRL;
+			op_entry->queued_ctrl.type = pkt_type;
+			op_entry->queued_ctrl.inject = inject;
+			dlist_insert_tail(&op_entry->queued_ctrl_entry,
+					  &ep->rx_entry_queued_ctrl_list);
 		}
 
-		return 0;
+		err = 0;
 	}
 
-	return rxr_pkt_post_ctrl_once(ep, entry_type, x_entry, ctrl_type, inject, flags);
+	return err;
 }
 
-ssize_t rxr_pkt_post_ctrl_or_queue(struct rxr_ep *ep, int entry_type, void *x_entry, int ctrl_type, bool inject)
+
+/**
+ * @brief post req packet(s). Queue the post if necessary
+ *
+ * Some REQ packet types must be queued.
+ *
+ * @param[in]   rxr_ep          endpoint
+ * @param[in]   x_entry         pointer to rxr_op_entry. (either a tx_entry or an rx_entry)
+ * @param[in]   pkt_type        packet type.
+ * @param[in]   inject          send control packet via inject or not.
+ * @return      On success return 0, otherwise return a negative libfabric error code. Typical error:
+ */
+ssize_t rxr_pkt_post_req(struct rxr_ep *ep, struct rxr_op_entry *tx_entry, int req_type, bool inject, uint64_t flags)
```

Comment:
what are the typical errors?

Reply:
Removed this not. There is no typical error for this function.